providers/oauth/oauth.httpobserver.js

```diff
@@ -75,12 +84,26 @@ FirefoxTabsAuth.prototype.launchAuthFlow = function (authUrl, stateObj, continua
     }, 100);
   };
 
+  fail = function () {
+    gBrowser.removeProgressListener(listener);
+    gBrowser.removeTab(tab);
+    continuation(undefined, 'Error in launchAuthFlow');
+  };
+
   httpRequestObserver.register();
 
   //Timeout before switching makes sure the http observer is installed.
   setTimeout(function () {
     tab = gBrowser.addTab(authUrl);
-    gBrowser.selectedTab = tab;
+    if (interactive) {
+      gBrowser.selectedTab = tab;
+    } else {
+      setTimeout(function () {
+        if (!hasCredentials) {
+          fail();
+        }
+      }, 5000);
```

Comment:
This allows time for the new tab to launch and be redirected from the Facebook URL to the direct URL we supply in the arguments.  I picked 5 seconds because it seemed like a reasonable amount of time to wait for a slower connection.  I could lower this if needed (that might make the integration tests easier if it were lowered below 3000ms), but I think it would be a pretty bad UX if the tab was closed too early causing the non-interactive login (reconnect in uProxy) to fail